### ISSUE TITLE
Exclude directories? 

### ISSUE THREAD

**Issue by Darkvater** [ncw/rclone#394]
How does one exclude directories (and their subdirectories)? I am trying to sync some folders that include subversion and git controlled checkout projects. My filter file has the following, which excluded all the files but keeps all the folders. What is the correct way? 

- \\.svn/**
- \\.git/**

**Comment by ncw** [comment]
That looks like the right syntax.

When you say keeps the folders - what do you mean exactly?  Currently rclone isn't very good at managing folders - see #100 for more info!

**Comment by Darkvater** [comment]
See attached images. With the above filter I have excluded the .git folders. This works, as in none of the files under .git and its subfolders are synced, however any directories with subdirectories are still transferred. This was a first-time copy to google drive so these files/folders did not exist anymore. #100 is more about deleting empty folders, but these folders should never have been synced in the first place as their parent (and grandparent, and grand-grandparent, etc.) are excluded.

Saying .git/** should exclude any path under .git. Not sure if this is a limitation of rclone of drive API?

![google_drive](https://cloud.githubusercontent.com/assets/2002972/13896817/a4aa7658-edd5-11e5-9e89-e31835453332.png)
![local_machine](https://cloud.githubusercontent.com/assets/2002972/13896822/c1a9a314-edd5-11e5-8203-f6df5b8784a3.png)

**Comment by ncw** [comment]
Ach, I should have noticed this earlier...

You need to use `/` when specifying filters not `\`, so the correct syntax is

```
/.svn/**
/.git/**
```

I'll write that in the docs.

**Comment by Darkvater** [comment]
Hi, thanks for the quick reply. Are you sure about that? .git and all subfiles/directories still get copied?

` 2016/03/20 20:32:02 Documents/programs/GitHub/minidlna/.git/objects/cc/b1a088ef3dade7b1264ce369f5eb69f971c6bf: Copied (new)
2016/03/20 20:32:02 Documents/programs/GitHub/minidlna/.git/objects/89/2d351365e1bea3b28b0286303eda059f5e8e56: Copied (new)
2016/03/20 20:32:02 Documents/programs/GitHub/minidlna/.git/objects/26/27a808e7a3808767ad58065f11c1fe309a7bca: Copied (new)`

My filter file is as follows:

```
tomi@alexandria:/mnt/sdb2$ cat ~/filter_file
- *.obj
- *.ncb
- /.svn/**
- /.git/**
- *.bak
- *.BAK
- *.suo
- *.idb
- *.ilk
- *.pdb
```
Using the following command:
```
tomi@alexandria:/mnt/sdb2$ work/rclone-v1.28-linux-arm/rclone --delete-excluded --filter-from ~/filter_file --dump-filters -v sync onedrive: google: 2>rclone_server_3.log
--- start filters ---
- (^|/)[^/]*\.obj$
- (^|/)[^/]*\.ncb$
- ^\.svn/.*$
- ^\.git/.*$
- (^|/)[^/]*\.bak$
- (^|/)[^/]*\.BAK$
- (^|/)[^/]*\.suo$
- (^|/)[^/]*\.idb$
- (^|/)[^/]*\.ilk$
- (^|/)[^/]*\.pdb$
--- end filters ---
```

**Comment by ncw** [comment]
The `/` at the start of the `/.svn/` anchors it to the root which is what I thought you wanted.

It looks like you actually want `.svn/**` which will match an `.svn` anywhere in a path component.

So I reckon this should be what you want!

```
- *.obj
- *.ncb
- .svn/**
- .git/**
- *.bak
- *.BAK
- *.suo
- *.idb
- *.ilk
- *.pdb
```

**Comment by Darkvater** [comment]
:) I see. Actually what I want is this below. I need to prefix it with a * otherwise it won't work. This also means the rsync and rclone filter is very similar 

```
- *.svn/**
- *.git/**
```

**Comment by ncw** [comment]
I tried the above without the initial `*` and it worked fine for me.  I tested it by doing this

```
$ cat /tmp/exludes 
- *.obj
- *.ncb
- .svn/**
- .git/**
- *.bak
- *.BAK
- *.suo
- *.idb
- *.ilk
- *.pdb
```

```
$ rclone ls --filter-from /tmp/exludes .
     3956 README.md
```

```
$ ls -lart
total 28
drwxr-xr-x 193 ncw ncw 16384 Mar 11 14:18 ..
-rw-rw-r--   1 ncw ncw  3956 Mar 11 14:20 README.md
drwxrwxr-x   3 ncw ncw  4096 Mar 11 14:20 .
drwxrwxr-x   8 ncw ncw  4096 Mar 11 14:21 .git
```

There is a ticket about recursing into excluded directories here #395 :-)

**Comment by Darkvater** [comment]
Let me try again, I might have fully gotten confused with the settings :(
Thanks for the prompt help!

**Comment by ncw** [comment]
I think this is all cleared up now, and I've written a bit more in the docs.

If not please feel free to re-open!

Thanks

Nick

**Comment by anthonyabraira** [comment]
I am on a Mac Server and tried to create an exclude-from file that contained these entries:

```
'*.'
'*.dmg'
'*.pkg'
'*.sparseimage'
'*.app'

```

However when I do a dry run of this in verbose mode, its still indicating some 503 errors on hidden folders and files. 

```
.TemporaryItems/folders.501: permission denied
RCHICAD 19 dmg/ArchiCAD 19/Add-Ons/Import-Export/Artlantis Render Studio 5 Out/Contents/Frameworks/CATL.framework/CATL: Can't transfer non file/directory
```

Is there something wrong with my exclusion file? Since the script and the exclusion file are in the same directory, I have setup like this: 

```
rclone/rclone sync --dry-run --exclude-from as_amazoncloud_excludes.txt  -v /path/rootfolder  as_amazoncloud:path/rootfolder
```

**Comment by dewi-ny-je** [comment]
Hi, this issue isn't actually solved in the current release 1.37 (tested on OmniOS/Illumos).

I have a filter "- .zfs/**" but still I get an error message:

```
2017/08/30 09:18:32 ERROR : olaf/.zfs/shares/olaf: Failed to copy: failed to open source object: open /tank/home/olaf/.zfs/shares/olaf: operation not applicable
```

That subfolder should not be checked at all!

The issue is very relevant on solaris: the folder .zfs contains snapshots of the whole dataset (think to a partition if you don't know ZFS), meaning that it contains the same files of the rest of the disk, as many times as the number of snapshots taken.

Other example with folder exclusion:
```
$ src/rclone-v1.37-solaris-amd64/rclone -v --fast-list --transfers 8 --exclude ".zfs/**" --exclude ".DS_Store" --exclude ".$EXTEND/**" sync /tank/eBooks/ secret:/tank/eBooks/
2017/08/30 20:42:05 INFO  : Encrypted drive 'secret:/tank/eBooks/': Modify window is 1ms
2017/08/30 20:42:08 INFO  : .$EXTEND/$QUOTA: Copied (new)
```
but here they are:
```
https://f001.backblazeb2.com/file/OmniOS/tank/eBooks/.$EXTEND/$QUOTA.bin
https://f001.backblazeb2.com/file/OmniOS/tank/eBooks/Scienze%2C+tecnica/Computer/.DS_Store.bin
```

**Comment by ncw** [comment]
@dewi-ny-je 

The `$EXTEND` excludes aren't working because the shell is expanding the `$EXTEND` first.  Use single quotes around the --exclude arguments to stop this.

I don't know why the `.zfs/**` exclude isn't working for you -  it seems to work for me.  Can you please make a [new issue on github](https://github.com/ncw/rclone/issues/new) with a script like the one below to reproduce the error please?

```
$ mkdir -p olaf/.zfs/shares/
$ touch olaf/.zfs/shares/olaf
$ chmod 000 olaf/.zfs/shares/olaf
$ rclone ls .
        0 olaf/.zfs/shares/olaf
$ rclone ls --filter '- .zfs/**' .
$ rclone cat .
2017/08/31 10:55:49 ERROR : olaf/.zfs/shares/olaf: Failed to open: open /tmp/extest/olaf/.zfs/shares/olaf: permission denied
$ rclone cat --filter '- .zfs/**' .
$ 
```

**Comment by dewi-ny-je** [comment]
Your example works, must have been a mistake on my side. I'll keep checking.
Thanks.
